contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,207 @@
+        if (currentReserveImbalance > 0) {
+            if (amount > currentReserveImbalance) {
+                emit CollateralAdded({amount: currentReserveImbalance, remaining: 0});
+                adjustedAmount = amount - currentReserveImbalance;
+                currentReserveImbalance = 0;
+            } else {
+                currentReserveImbalance -= amount;
+                emit CollateralAdded({amount: amount, remaining: currentReserveImbalance});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        totalMinted += adjustedAmount;
+        emit NativeTokensMinted(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-transferToSource}.
+     */
+    function transferToSource(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero Recipient Address"
+        );
+
+        require(
+            currentReserveImbalance == 0,
+            "NativeTokenDestination: Cannot release tokens until source contract is collateralized"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedFeeAmount = 0;
+        if (feeInfo.amount > 0) {
+            adjustedFeeAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeInfo.contractAddress),
+                feeInfo.amount
+            );
+            IERC20(feeInfo.contractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        // Burn native token by sending to BLACKHOLE_ADDRESS
+        payable(BLACKHOLE_ADDRESS).transfer(msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToSource({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    function isCollateralized() external view returns (bool) {
+        return currentReserveImbalance == 0;
```

Comment:
Same comment here, may want to consider changing the condition to `currentReserveImbalance <= 0`

Reply:
Same as above